utils/rf.py

```diff
@@ -0,0 +1,368 @@
+    3) Returns confusion matrix as a list of struct
+
+    :param Table ht: Input table
+    :param PipelineModel rf_model: RF Model
+    :param list of str features: Columns containing features that were used in the model
+    :param str label: Column containing label to be predicted
+    :param str prediction_col_name: Where to store the prediction
+    :return: A list containing structs with {label, prediction, n}
+    :rtype: list of Struct
+    """
+
+    ht = apply_rf_model(ht.filter(hl.is_defined(ht[label])),
+                        rf_model,
+                        features,
+                        label,
+                        prediction_col_name=prediction_col_name)
+
+    test_results = ht.group_by(ht[prediction_col_name], ht[label]).aggregate(n=hl.agg.count()).collect()
+
+    #Print results
+    df = pd.DataFrame(test_results)
+    df = df.pivot(index=label, columns=prediction_col_name, values='n')
+    logger.info("Testing results:\n{}".format(pformat(df)))
+    logger.info("Accuracy: {}".format(
+        sum([x.n for x in test_results if x[label] == x['rf_prediction']]) /
+        sum([x.n for x in test_results])
+    ))
+
+    return test_results
+
+
+
+def apply_rf_model(
+        ht: hl.Table,
+        rf_model: pyspark.ml.PipelineModel,
+        features: List[str],
+        label: str,
+        probability_col_name: str = 'rf_probability',
+        prediction_col_name: str = 'rf_prediction'
+) -> hl.Table:
+    """
+    Applies a Random Forest (RF) pipeline model to a Table and annotate the RF probabilities and predictions.
+
+    :param MatrixTable ht: Input HT
+    :param PipelineModel rf_model: Random Forest pipeline model
+    :param list of str features: List of feature columns in the pipeline. !Should match the model list of features!
+    :param str label: Column containing the labels. !Should match the model labels!
+    :param str probability_col_name: Name of the column that will store the RF probabilities
+    :param str prediction_col_name: Name of the column that will store the RF predictions
+    :return: Table with RF columns
+    :rtype: Table
+    """
+
+    logger.info("Applying RF model.")
+
+    has_idx = 'idx' in ht.row
+    if not has_idx:
+        ht = ht.add_index()
+
+    df = ht_to_rf_df(ht, features, label, 'idx')
+
+    rf_df = rf_model.transform(df)
+
+    rf_ht = hl.Table.from_spark(
+        rf_df.rdd.map(
+            lambda row:
+            Row(idx=row['idx'],
+                probability=row["probability"].toArray().tolist(),
+                predictedLabel=row["predictedLabel"])
+        ).toDF()
+    ).persist()
+
+    rf_ht = rf_ht.key_by('idx')
+
+    ht = ht.annotate(
+        **{
+            probability_col_name: {label: rf_ht[ht['idx']]["probability"][i] for i, label in enumerate(get_labels(rf_model))},
+            prediction_col_name: rf_ht[ht['idx']]["predictedLabel"]
+        }
+    )
+
+    if not has_idx:
+        ht = ht.drop('idx')
```

Comment:
Can't drop a key. I think you now would need to `ht = ht.key_by().drop('idx')`

Reply:
Decided to always index rather than rely on previously annotated `idx` column if any. Seems more robust and flow makes more sense now I think.